### PR TITLE
Create EC2 resource using terraform

### PR DESCRIPTION
**Pull request by guy9050**
This PR will contain a terraform plan for creating an EC2 instace for hosting CRC.

Signed-off-by: Guy Inger <ginger@redhat.com>

### CODE REVIEW

**Comment by guy9050** [comment]
I was able to successfully create and destroy an EC2 instance with this terraform file in this PR.

Diving into the scrip though, [it does more things](https://github.com/crc-org/crc-cloud/blob/main/crc-cloud.sh#L201) rather than just create an EC2 instance.
For example, the [cluster installation step](https://github.com/crc-org/crc-cloud/blob/main/crc-cloud.sh#L132), do we want to it with terraform? Or would it make sense to integrate Ansible and use it for installing the cluster, [swapping the ssh_keys](https://github.com/crc-org/crc-cloud/blob/main/crc-cloud.sh#L117), i.e all the things that are not related to the EC2 resource itself?

**Comment by lmilbaum** [comment]
> I was able to successfully create and destroy an EC2 instance with this terraform file in this PR.
> 
> Diving into the scrip though, [it does more things](https://github.com/crc-org/crc-cloud/blob/main/crc-cloud.sh#L201) rather than just create an EC2 instance. For example, the [cluster installation step](https://github.com/crc-org/crc-cloud/blob/main/crc-cloud.sh#L132), do we want to it with terraform? Or would it make sense to integrate Ansible and use it for installing the cluster, [swapping the ssh_keys](https://github.com/crc-org/crc-cloud/blob/main/crc-cloud.sh#L117), i.e all the things that are not related to the EC2 resource itself?

I don't think the cluster installation should be handled by Terraform.  I think it is part of the provisioning phase. Ansible is better suited for such a task.
Regarding the key swapping. I don't understand why it is needed.  Terraform can create a key pair and place the pem file in the developers ~/.ssh folder.  Feel free to poke into what Yael has done in `coffr/poc`.

**Comment by adrianriobo** [comment]
There are several question around this: 

* About the setup of the cluster, currently this is done based on the bash scripts, but there is a spike on going to add the logic to the bundle itself so basically it would be only required to pass certain param when booting the machine and the bundle will be self configured, so not sure how much to we want to deal with it...actually this kind of "conflicts" with some points on #23 probably this should be mention during next meeting.

* About the swapping, there are certain requirements that providers would ask on the imported images. at the moment I can not remember if the key setup on AWS is done with an agent or based on cloud-init, anyway images from crc does not include such solutions , so even when you create a aws keypair it is not setup on the machine as the user key... this is done with the key swapping script mentioned by @guy9050

**Comment by gbraad** [comment]
>  I am working on the pulumi PoC and this is how I am doing it there

@tsebastiani QE has been using a similar tool for their own testing purposes, and this is what Adrian has been working on for some time. It does the same as `crc-cloud.sh`. Let's also have a look at his code/PoC and see how we can adopt ideas or patterns, as I agree, maintainability is key.

One of the preparations on our end for the 'key swapping' and other tasks it to make the images more self-container/self-sufficient: https://github.com/crc-org/snc/issues/638. Ideally we make modifications that work for both CRC and CRC-Cloud. I am not sure if Ansible is a necessity, but could be considered for more advanced tasks is needed after the main provisioning, though should not be a hard requirement as this might impact CRC.

**Comment by tsebastiani** [comment]
> 

AWS keypair creation is done only because is mandatory, in order to make it work (as you said) the image should include some tweaks that are not actually implemented in the image (like cloud-init). To overcome this issue, and make the newly deployed image safe, a keypair is generated by the script and is swapped with the default one.

**Comment by lmilbaum** [comment]
Several points were mentioned in the discussion. I want to focus on one of them. This PR is about bootstrapping only. It deals with only the AWS resources. Key Pair is one of those resources. AFAIK, there are two spikes worked in parallel. One is done one and the second one is with Pulumi.  Eventually, we would pick one of them and drop the other. This was my understanding.

File: terraform/aws/aws.tf [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  default = "ami-0569ce8a44f2351be"
+}
+
+variable "instance_type" {
+  type = string
+  default = "c6in.2xlarge"
+}
+
+
+# Specify the cloud provider
```
Comment (guy9050) [inline terraform/aws/aws.tf:25]:
```suggestion
```

File: terraform/aws/aws.tf [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+resource "aws_instance" "web" {
+  ami           = var.ami
+  instance_type = var.instance_type
+  security_groups = ["ssh", "https", "api"]
+  tags = {
+    Name = "web"
+  }
+}
+
+# Create a resource of type "aws_security_group" to enable SSH
```
Comment (guy9050) [inline terraform/aws/aws.tf:42]:
```suggestion

```

**Comment by guy9050** [comment]
Not relevant anymore.

**Pull request closed without merging**